### PR TITLE
feat(scheduler-targets): CodeBuild scheduler target

### LINKED ISSUE #27448: (aws-scheduler-targets-alpha): Add CodeBuildStartBuild Target
### Describe the feature

Work to support L2 constructs for AWS Scheduler is in progress (https://github.com/aws/aws-cdk/issues/23394). See the approved [RFC](https://github.com/aws/aws-cdk-rfcs/pull/486/files). RFC planned to add 12 templates targets, but only Lambda Invoke is currently implemented (https://github.com/aws/aws-cdk/pull/26575). 

This issue tracks implementation of  `CodeBuildStartBuild` target to start an AWS CodeBuild build. 

### Use Case

Customers would like to use templated target `CodeBuildStartBuild` to be able to start AWS CodeBuild build on schedule. L2 target construct  should grant required permissions to the `scheduler` to start the CodeBuild.  

### Proposed Solution

The proposed solution needs to be adopted to the recent examples of `LambdaInvoke` (https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-scheduler-targets-alpha/lib/lambda-invoke.ts). 

Solution should also include unit and integration tests. 

```ts
export class CodeBuildStartBuild extends ScheduleTargetBase implements IScheduleTarget {
    constructor(
      private readonly project: codebuild.IProject,
      private readonly props: ScheduleTargetBaseProps,
    ) {
      super(props, project.projectArn)
    }

    protected addTargetActionToRole(schedule: ISchedule, role: IRole): void {
      // TODO: Check if target and schedule are in the same account and region
      // TODO: Check if target and role are in the same account

     // TODO: Check if L2 grant methods can be used
      role.addToPrincipalPolicy(new iam.PolicyStatement({
        actions: ['codebuild:StartBuild'],
        resources: [this.project.projectArn],
      }));
    }
  }
```

### Other Information

_No response_

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.99.1

### Environment details (OS name and version, etc.)

MacOs

### PR DESCRIPTION
**Pull request by daschaa**
Closes #27448.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-scheduler-targets-alpha/README.md [hunk 1]
```diff
@@ -102,3 +102,21 @@ new Schedule(this, 'Schedule', {
   }),
 });
 ```
+
+## Start a CodeBuild job
```
Comment (lpizzinidev) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/README.md:106]: Can you please add `CodeBuildStartBuild` to the [list of supported targets](https://github.com/aws/aws-cdk/blob/576bb2af0dffe88d35ea370ee8ef07f3d12872f0/packages/%40aws-cdk/aws-scheduler-targets-alpha/README.md?plain=1#L25-L28)?

File: packages/@aws-cdk/aws-scheduler-targets-alpha/test/codebuild-start-build.test.ts [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+
+  beforeEach(() => {
+    app = new App();
+    stack = new Stack(app, 'Stack', { env: { region: 'us-east-1', account: '123456789012' } });
+    codebuildProject = new Project(stack, 'Project', {
+      buildSpec: BuildSpec.fromObject({}),
+    });
+  });
+
+  test('creates IAM role and IAM policy for step function target in the same account', () => {
```
Comment (lpizzinidev) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/test/codebuild-start-build.test.ts:26]:
```suggestion
  test('creates IAM role and IAM policy for code build target in the same account', () => {
```

File: packages/@aws-cdk/aws-scheduler-targets-alpha/test/codebuild-start-build.test.ts [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+            Effect: 'Allow',
+            Resource: codebuildArnRef,
+          },
+        ],
+      },
+      Roles: [{ Ref: 'SchedulerRoleForTarget1441a743A31888' }],
+    }, 1);
+  });
+
+  test('creates IAM policy for imported state machine in the same account', () => {
```
Comment (lpizzinidev) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/test/codebuild-start-build.test.ts:161]:
```suggestion
  test('creates IAM policy for imported code build project in the same account', () => {
```

File: packages/@aws-cdk/aws-scheduler-targets-alpha/test/codebuild-start-build.test.ts [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+    const codeBuildTarget = new CodeBuildStartBuild(importedProject, {});
+
+    expect(() =>
+      new Schedule(stack, 'MyScheduleDummy', {
+        schedule: expr,
+        target: codeBuildTarget,
+      })).toThrow(expectedError);
+  });
+
+  /**/
```
Comment (lpizzinidev) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/test/codebuild-start-build.test.ts:291]:
```suggestion
```

**Review by lpizzinidev — CHANGES_REQUESTED** [review]
Looks good to me, just some minor adjustments to be made.

File: packages/@aws-cdk/aws-scheduler-targets-alpha/README.md [hunk 5]
```diff
@@ -26,6 +26,7 @@ The following targets are supported:
 
 1. `targets.LambdaInvoke`: [Invoke an AWS Lambda function](#invoke-a-lambda-function))
 2. `targets.StepFunctionsStartExecution`: [Start an AWS Step Function](#start-an-aws-step-function)
+2. `targets.CodeBuildStartBuild`: [Start a CodeBuild job](#start-a-codebuild-job)
```
Comment (lpizzinidev) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/README.md:29]:
```suggestion
3. `targets.CodeBuildStartBuild`: [Start a CodeBuild job](#start-a-codebuild-job)
```

File: packages/@aws-cdk/aws-scheduler-targets-alpha/lib/codebuild-start-build.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import { ScheduleTargetBase, ScheduleTargetBaseProps } from './target';
+import { sameEnvDimension } from './util';
+
+/**
+ * Use an AWS CodeBuild as a target for AWS EventBridge Scheduler.
+ */
+export class CodeBuildStartBuild extends ScheduleTargetBase implements IScheduleTarget {
+  constructor(
+    private readonly project: IProject,
+    private readonly props: ScheduleTargetBaseProps,
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/lib/codebuild-start-build.ts:14]:
```suggestion
    private readonly props: ScheduleTargetBaseProps = {},
```

File: packages/@aws-cdk/aws-scheduler-targets-alpha/README.md [hunk 7]
```diff
@@ -102,3 +103,21 @@ new Schedule(this, 'Schedule', {
...
+called every hour by Event Bridge Scheduler.
+
+```ts
+import * as codebuild from 'aws-cdk-lib/aws-codebuild';
+
+declare const project: codebuild.Project;
+
+new Schedule(this, 'Schedule', {
+  schedule: ScheduleExpression.rate(Duration.minutes(60)),
+  target: new targets.CodeBuildStartBuild(project, {}),
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/README.md:121]:
```suggestion
  target: new targets.CodeBuildStartBuild(project),
```

**Review by lpizzinidev — APPROVED** [review]

File: packages/@aws-cdk/aws-scheduler-targets-alpha/lib/codebuild-start-build.ts (hunk 6 above)
Comment (kaizencc) [reply packages/@aws-cdk/aws-scheduler-targets-alpha/lib/codebuild-start-build.ts:14]: an aside: why did we call this props if they were all optional lol

**Review by kaizencc — DISMISSED** [review]
Looking great @daschaa. Minor comment

**Comment by lpizzinidev** [comment]
@daschaa 
Also, the module in the title should not contain `alpha`: `feat(scheduler-targets) ...`

File: packages/@aws-cdk/aws-scheduler-targets-alpha/lib/codebuild-start-build.ts (hunk 6 above)
Comment (daschaa) [reply packages/@aws-cdk/aws-scheduler-targets-alpha/lib/codebuild-start-build.ts]: aaah yes. a default value totally makes sense. thanks for pointing out haha.

**Review by kaizencc — APPROVED** [review]
Thanks @daschaa! Thanks for the reviews @lpizzinidev!

**Pull request merged**